# HACK-1. Яндекс.Еда 
## Решение группы №8

## Шаг 0. Импортируем библиоиеки

In [76]:
import pandas as pd

## Шаг 1. Загрузм данные

In [6]:
df = pd.read_excel('https://lms.skillfactory.ru/assets/courseware/v1/2222decafe3827199aaac54c7c0e3569/asset-v1:SkillFactory+DST-HACK+30OCT2020+type@asset+block/yandex_hack.xlsx')

## Шаг 2. Посмотрим на представленный датасет

In [76]:
df.sample(10)

order_created_datetime brand_name  \
55334  2015-09-14 15:32:24.000000          B   
13939  2015-09-14 09:32:04.000000          B   
38102  2015-09-14 09:10:01.000000          B   
45251  2015-09-13 19:12:14.000000          B   
24966  2015-09-16 10:01:55.000000          B   
25298  2015-09-13 16:55:39.000000          B   
27456  2015-09-15 20:35:05.000000          B   
44453  2015-09-13 11:40:49.000000          B   
36298  2015-09-14 15:47:58.000000          A   
14961  2015-09-15 20:55:41.000000          B   

                                rest_id batched_with_order_id  \
55334  6ac1b8fed2910f79982117e6cd3c3ad3                   NaN   
13939  16a7bc2da1ef26f1fb9221c155a4a8f2                   NaN   
38102  88868c98beb65cedd2e073869e0ca44e                   NaN   
45251  19db767f8a848d622b9b1ee9cfa558f0                   NaN   
24966  9b06c08e9af17ccc5736271c88cfb790                   NaN   
25298  837603ca10bc2e18a83aec81bea81aab                   NaN   
27456  b4f00726ee2d79fe7fb0ef6d317e6d15                   NaN   
44453  7cc98331ddb8ccffe464455f3df763d9                   NaN   
36298  12da84be1fbccae286c895c3aaf616e7                   NaN   
14961  f9e4306b064810d3fd3b4846fc5313b6                   NaN   

                               order_id  first_in_multiorder_flg  \
55334  c0297cfd90c93d215e12710eb3ca2075                      NaN   
13939  159cc4cbba2ca6a369da7b8ee6392287                      NaN   
38102  5e6931317d8b77aef4bc94faa1ac9164                      NaN   
45251  71f6c8fa88ea253dc7d500562459b266                      NaN   
24966  7cbf655c5ae962ff0db6ed52ac055d3b                      NaN   
25298  f2e214b43c2e192afc18d15a118357d7                      NaN   
27456  aee5b8ccc16c4deae1fa2f3dc0bccaf9                      NaN   
44453  85b4d657bc09616e16a4de0ecd8ffdf1                      NaN   
36298  b21681e22460f72154e70880e5143ad2                      NaN   
14961  c23c15ef693590cfc62e8d59ab754369                      NaN   

                             courier_id  order_items_cnt  \
55334  c2f92caeda4e57703bd9b8b9fb5a9418                1   
13939  455c17bdc5f3258cf42ac0372df5b71f                3   
38102  1a7cff01707d9946440f2787cfb74afa                6   
45251  a6a3fe205e73b553fb2eabc43bf62b48                5   
24966  44cf34f43cbbcb4ed4553d483be40c10               10   
25298  d9029e0f9e71194979cb76411cd145a7                2   
27456  c7829cc1666b00b3b84b4757a9decea6               11   
44453  2f7a2637242f54d9c78427203fdb962e                3   
36298  a2687134e8d379940b8a77ce44d76518                8   
14961  08099bab6f63d2580b4578d166fab7eb                5   

                                   city  
55334  4805ee58a2d975322bda9710e2116a17  
13939  e6705560085199e9674cd721ee844d0d  
38102  e6705560085199e9674cd721ee844d0d  
45251  e6705560085199e9674cd721ee844d0d  
24966  8ef4fd9e60ef979b27a7ce3487dce62d  
25298  e6705560085199e9674cd721ee844d0d  
27456  e6705560085199e9674cd721ee844d0d  
44453  c6fbacc5cab7e134c9db2a4bbfcc5b64  
36298  29b5d0e91de8f84cbb829977dfb8f4a3  
14961  ea9f1e9302826b93f5ac1898375fbc38

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66464 entries, 0 to 66463
Data columns (total 9 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   order_created_datetime   66464 non-null  object 
 1   brand_name               66464 non-null  object 
 2   rest_id                  66464 non-null  object 
 3   batched_with_order_id    442 non-null    object 
 4   order_id                 66464 non-null  object 
 5   first_in_multiorder_flg  442 non-null    float64
 6   courier_id               66202 non-null  object 
 7   order_items_cnt          66464 non-null  int64  
 8   city                     66464 non-null  object 
dtypes: float64(1), int64(1), object(7)
memory usage: 4.6+ MB


### В датасете 9 столбцов. Из них:
* 2 числовых
* 7 текстовых

В данных пости нет пропусков за исключеним столбцов "first_in_multiorder_flg" и "order_items_cnt" - в них по 442 записи. Возможно это - обучающая выборка. 

В столбце "courier_id" тоже есть несколько пропксков. Но, судя по названию это номер курьера - можно заменить любым значением.

## Шаг 3. Посмотрим на наши столбцы более пристально

### Шаг 3.1 order_created_datetime
В столбце записано время создания заказа.
Формат такстовый переведем в datetime

In [78]:
df.order_created_datetime.min()

'2015-09-13 00:00:05.000000'

In [79]:
df.order_created_datetime.max()

'2015-09-16 10:56:59.000000'

В столбце данные за три с половиной дня.

С 13 сентября ВОСКРЕСЕНЬЯ по 16 сентября СРЕДУ 2015 года. 

Выходные попадают воскресеньем. Может это надо будет учесть в дальнейшем.

UPD. Данные изменены, так что, не смотрим на дни недели.

### Шаг 3.2 brand_name

In [80]:
df.brand_name.value_counts()

B    56835
A     9629
Name: brand_name, dtype: int64

Ага. Тут два значения A и B.
Видимо по ним отберем обучающую выборку.

## Шаг 3.3 rest_id

In [81]:
df.rest_id.value_counts()

7cc98331ddb8ccffe464455f3df763d9    606
c990b25894426c0ac4767680c5951ab2    604
cdfe9bf9a7d4f3a8d6fed9d032e26e7e    511
0d12507f6de6125f6288915c7dd06783    482
d9e5f2bf350b61f1c2cc15f044bda13d    477
                                   ... 
e669eb8e583c5cf61df563b189f978bb      1
58362b85407562242fe2cb66cde31c9a      1
dafa9414bfd7c5d724966213b4ca38c9      1
0681bec9a7c69c0053a30e160b568240      1
9defb7ad40926d020419ba2b1c393bd1      1
Name: rest_id, Length: 1098, dtype: int64

У нас в выбоке 1098 уникальных ресторана

## Шаг 3.4 batched_with_order_id

In [83]:
df[df.brand_name == 'A'].batched_with_order_id.value_counts()

095c886613a6480c8bbe5e32c48a146d    1
095c74f8edd6dc3e8dce39e8f325a908    1
7c8e9cab946ee6235b60af8101d66fd3    1
f44e1d45a5c56382f187020506f3e19b    1
949211dd9049bbfa4e66bae30cc1b859    1
                                   ..
c05ac5763ca7efadde532f2930dc8963    1
38114f43b317b925c2847f8d79e41d3f    1
e0afdf0b2bc0f617a3181d7d5bacf9be    1
d8230247eb6a7ce1c0fbd0217b0224d2    1
4d225c3fbdf5a80191a5c842f3a2f848    1
Name: batched_with_order_id, Length: 442, dtype: int64

## Шаг 3.5 order_id

In [84]:
df.order_id.value_counts()

09d73892c20580b092fdd2059bc334c4    1
2351a1284143cd466e1870bc0c1de889    1
e3d0b5edc3aff1d2f58d55630a3dffae    1
160bede0d0f8403ff9cfc7917f5284a4    1
8d9628482c034bfb77623571dae7e944    1
                                   ..
73d1bc528d8c5368b2fe86b670b67f71    1
5b608de3bf2b7e1c4db8d77d457c7b89    1
d37faa50e2b7a70515961106f0a55a9c    1
8b87e750b9b85696e15afa40aa6b45f6    1
c98cc367614ee855585b5973d6a8fee9    1
Name: order_id, Length: 66464, dtype: int64

In [85]:
df[df.batched_with_order_id.notna() == True]

order_created_datetime brand_name  \
257    2015-09-13 17:37:08.000000          A   
435    2015-09-13 11:18:49.000000          A   
526    2015-09-15 11:55:03.000000          A   
684    2015-09-15 17:36:03.000000          A   
717    2015-09-13 17:08:07.000000          A   
...                           ...        ...   
65377  2015-09-14 14:40:35.000000          A   
65416  2015-09-14 14:35:20.000000          A   
65456  2015-09-15 16:35:20.000000          A   
65595  2015-09-14 23:25:22.000000          A   
65893  2015-09-13 16:10:19.000000          A   

                                rest_id             batched_with_order_id  \
257    8da96a16838351f6f096d45a4e159f4d  89164b34e9d20a1c1918d156ca7c2b1e   
435    b3bf2fc20c4646298c05716fbe1aa135  afc4c403cc654734538b007fdbc87664   
526    2acb4a3db77b139dda1cd6bcbbfd4b8e  986941e665d646910d432b5f48f9b94c   
684    525505123c94df09ef5b5614e4afe7b2  d469fe73826b9b3f1a3ee58869d91757   
717    680fee36fb26805d3bf7af52cdb44f0f  831b2771564b022030ce699e6807ab03   
...                                 ...                               ...   
65377  974d115d8040e56fcac9638bd33e1731  bdb29739e550cd9bce8d0b2e07b6f33b   
65416  974d115d8040e56fcac9638bd33e1731  759928f3b5075cf0eea68cc97d17cf06   
65456  b3bf2fc20c4646298c05716fbe1aa135  f67eb84cf826ea33988019b56cf99b6b   
65595  b3bf2fc20c4646298c05716fbe1aa135  0e0984a5a2019591d54af8bde172e75b   
65893  485498fd82fb0d82fd89da0b8af280d4  472daa3056f5a59dc94f86290390e802   

                               order_id  first_in_multiorder_flg  \
257    9f8fb710b29202ac3dc36a6712186de4                      0.0   
435    d887b9731cf06e5dd77623ef5ad56cd8                      0.0   
526    fd0dde82caff2d709ac803c8872fd1e9                      0.0   
684    b4e39b3e84e81f1ea34c353896ff6853                      0.0   
717    6efbee90330a0bced1f7938a2e209054                      0.0   
...                                 ...                      ...   
65377  759928f3b5075cf0eea68cc97d17cf06                      1.0   
65416  bdb29739e550cd9bce8d0b2e07b6f33b                      0.0   
65456  517d0dcde489a94e594d2c028dbcb84a                      1.0   
65595  a91a73cc77884c0c594d0a600aa0fe5b                      0.0   
65893  570125290d88158a8e02b7663f09363e                      0.0   

                             courier_id  order_items_cnt  \
257    2644231e976679c41b15a11524a4e530                8   
435    8632dc891885aba8e572522dedc7572f                3   
526    7e8de661602e618fe8cb16c8b3e02a95                3   
684    876e5276f463736bf52fc540664fd02e                5   
717    8a2d57b71452eed5c78c3f81ee2999d2                3   
...                                 ...              ...   
65377  88750b500f3f628d7d85b352f6113736                2   
65416  88750b500f3f628d7d85b352f6113736                7   
65456  310b2378b5bf2b9f7046c70c3082103d                3   
65595  d41ff061323280cf0cebd9adb7a0755e                5   
65893  3f3c902e25018700cacecf38f5414e45                4   

                                   city  
257    e6705560085199e9674cd721ee844d0d  
435    e6705560085199e9674cd721ee844d0d  
526    e6705560085199e9674cd721ee844d0d  
684    e6705560085199e9674cd721ee844d0d  
717    e6705560085199e9674cd721ee844d0d  
...                                 ...  
65377  e6705560085199e9674cd721ee844d0d  
65416  e6705560085199e9674cd721ee844d0d  
65456  e6705560085199e9674cd721ee844d0d  
65595  e6705560085199e9674cd721ee844d0d  
65893  e6705560085199e9674cd721ee844d0d  

[442 rows x 9 columns]

мульти заказов всего 442

In [86]:
df[df.batched_with_order_id.notna() == True].first_in_multiorder_flg.value_counts()

1.0    221
0.0    221
Name: first_in_multiorder_flg, dtype: int64

### 3.7 order_items_cnt 

In [89]:
df[df.batched_with_order_id.notna() == True].order_items_cnt.value_counts()

3     96
4     79
5     59
2     58
6     37
7     35
8     20
1     18
9     11
10     8
12     5
13     5
11     4
15     4
16     1
14     1
22     1
Name: order_items_cnt, dtype: int64

In [10]:
df['DateTime'] = pd.to_datetime(df['order_created_datetime'], format='%Y-%m-%d %H:%M:%S.%f') #%S%d%b%Y:%H:%M:%S.%f')

In [11]:
df = df.sort_values(by=['DateTime'])

In [103]:
df

order_created_datetime brand_name  \
64367  2015-09-13 00:00:05.000000          A   
58676  2015-09-13 00:00:13.000000          A   
11760  2015-09-13 00:00:39.000000          A   
41675  2015-09-13 00:00:44.000000          B   
41533  2015-09-13 00:00:53.000000          B   
...                           ...        ...   
62680  2015-09-16 10:56:45.000000          B   
36687  2015-09-16 10:56:48.000000          A   
43883  2015-09-16 10:56:53.000000          B   
2612   2015-09-16 10:56:57.000000          B   
65498  2015-09-16 10:56:59.000000          B   

                                rest_id batched_with_order_id  \
64367  d1e66c75e8deec66a0aa08aefa5736aa                   NaN   
58676  39baa0eae83853d736e398093239dcd5                   NaN   
11760  d1e66c75e8deec66a0aa08aefa5736aa                   NaN   
41675  69a08198a23fac8f83359e2ec23a49b6                   NaN   
41533  69a08198a23fac8f83359e2ec23a49b6                   NaN   
...                                 ...                   ...   
62680  0596c848dec6a3c13396cc4b80f8974c                   NaN   
36687  12da84be1fbccae286c895c3aaf616e7                   NaN   
43883  7cc98331ddb8ccffe464455f3df763d9                   NaN   
2612   ce43eec08c9089152e6a8d6f2b6cc888                   NaN   
65498  0437d5ccc002a15c19d89fd98c80035c                   NaN   

                               order_id  first_in_multiorder_flg  \
64367  ffc6ecf37c8391821bda9111320f299c                      NaN   
58676  49d061c7882d839710443d8ac8108d7e                      NaN   
11760  3e293d204168757552ebd7b62ea38889                      NaN   
41675  d9f50562ab79c432e64413fcee7aeb9c                      NaN   
41533  5cc2b1699fdbeee506fb770ff77c4ff9                      NaN   
...                                 ...                      ...   
62680  4d79146812797a87907c6f5b144596a9                      NaN   
36687  aaf3bdc003bbc8dd60b8bad2045f700a                      NaN   
43883  06565a86181e342c7d64b1ea788b308c                      NaN   
2612   de69eb79833a6f417f316ab70c17ce9c                      NaN   
65498  6205554e043f0d84544ef6d22eae9ad2                      NaN   

                             courier_id  order_items_cnt  \
64367  d016782f2092ef2b2f943e736ececced                2   
58676  5b17547f825d9cf35c7853621dd842f8                7   
11760  d016782f2092ef2b2f943e736ececced               10   
41675  8ec5e3cc3bc59b70a0c8863b0355cc92                3   
41533  50c20cfed81be8a8fbcd8dc5b2e2ab76                7   
...                                 ...              ...   
62680  4c122524cdec8e402db29fb2e7136508                9   
36687  792e39d53e4317e13a99eb8a9bec19f9                2   
43883  906808228a91832545f6e52a6c202739                2   
2612   6d684c83727f7ca26516270d56d6d1f9                3   
65498  86a3494da03e99f76c8fba70d92173c4                1   

                                   city            DateTime  
64367  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:05  
58676  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:13  
11760  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:39  
41675  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:44  
41533  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:53  
...                                 ...                 ...  
62680  29b5d0e91de8f84cbb829977dfb8f4a3 2015-09-16 10:56:45  
36687  29b5d0e91de8f84cbb829977dfb8f4a3 2015-09-16 10:56:48  
43883  c6fbacc5cab7e134c9db2a4bbfcc5b64 2015-09-16 10:56:53  
2612   feb10fbf40476b6ca4793a84b67011cf 2015-09-16 10:56:57  
65498  e6705560085199e9674cd721ee844d0d 2015-09-16 10:56:59  

[66464 rows x 10 columns]

Создаём новые ДатаФреймы

In [14]:
dfA_M = df[df.batched_with_order_id.notna() == True]

In [33]:
dfA = df[df.brand_name == 'A']

In [77]:
dfB = df[df.brand_name == 'B']

In [191]:
A = len(dfA)
A_M = len(dfA_M)
percent = (A_M/2) / (A - A_M/2)
print("Процент двойных заказов от общего числа заказов", round(percent*100, 2),'%')

Процент двойных заказов от общего числа заказов 2.35 %


## Шаг 5. Посмотрим на время заказов

In [192]:
# Посчитаем разницу в сикундах между соседними заказами

import numpy as np
u = []
a = -1000
for i in dfA['DateTime']:
  if a == -1000:
    a = i
    u.append(0)
    continue
  td = i - a
  u.append((td / np.timedelta64(1, 's')))
  a = i

In [193]:
print('Среднее время меджу заказами в выборке А:', sum(u)/len(u))

Среднее время меджу заказами в выборке А: 31.01080070620002


In [194]:
dfA.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9629 entries, 64367 to 36687
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   order_created_datetime   9629 non-null   object        
 1   brand_name               9629 non-null   object        
 2   rest_id                  9629 non-null   object        
 3   batched_with_order_id    442 non-null    object        
 4   order_id                 9629 non-null   object        
 5   first_in_multiorder_flg  442 non-null    float64       
 6   courier_id               9437 non-null   object        
 7   order_items_cnt          9629 non-null   int64         
 8   city                     9629 non-null   object        
 9   DateTime                 9629 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(1), object(7)
memory usage: 827.5+ KB


In [198]:
dfA_M.head()

order_created_datetime brand_name  \
52465  2015-09-13 04:00:11.000000          A   
25870  2015-09-13 09:11:50.000000          A   
25804  2015-09-13 09:12:49.000000          A   
23945  2015-09-13 09:53:28.000000          A   
23701  2015-09-13 10:00:37.000000          A   

                                rest_id             batched_with_order_id  \
52465  6c1ad5c66540b8c174ff7d9307027299  d0febfc6653fa848cf0a9fadf98261dd   
25870  8aec6c26cf9db4db10ed06e5e39681ed  04c7d89ab6b8ff72fb5868413602ddef   
25804  8aec6c26cf9db4db10ed06e5e39681ed  5ae2e25193da2000e07dc2051b5ca657   
23945  707f1da36335ef7234721b49991d39c3  83044987a923b13056db64220dd548c5   
23701  707f1da36335ef7234721b49991d39c3  287564b3ea79fef51a657c6a4237deb5   

                               order_id  first_in_multiorder_flg  \
52465  931ad84269827a4c27ce66734460cdf7                      0.0   
25870  5ae2e25193da2000e07dc2051b5ca657                      1.0   
25804  04c7d89ab6b8ff72fb5868413602ddef                      0.0   
23945  287564b3ea79fef51a657c6a4237deb5                      0.0   
23701  83044987a923b13056db64220dd548c5                      1.0   

                             courier_id  order_items_cnt  \
52465  a973a8d1969c72a9eb8147d995e519bf                7   
25870  ebe67e247939ab69f7734ba1557b763d                2   
25804  ebe67e247939ab69f7734ba1557b763d               14   
23945  cd30a21c6bbbcc784de70aa8627be584                8   
23701  cd30a21c6bbbcc784de70aa8627be584                5   

                                   city            DateTime    UNX_time  
52465  e6705560085199e9674cd721ee844d0d 2015-09-13 04:00:11  1442116811  
25870  e6705560085199e9674cd721ee844d0d 2015-09-13 09:11:50  1442135510  
25804  e6705560085199e9674cd721ee844d0d 2015-09-13 09:12:49  1442135569  
23945  e6705560085199e9674cd721ee844d0d 2015-09-13 09:53:28  1442138008  
23701  e6705560085199e9674cd721ee844d0d 2015-09-13 10:00:37  1442138437

На данном этапе создаю новый столбец с датой в формате UNIX. Это количество секунд, прошедших с полуночи 1 января 1970 года по часовой зоне UTC. Преимущество данного формата в его простоте: обычно это целое число. Также им удобно считать разницу между двумя датами: это будет разница двух времен в секундах.

In [18]:
# df['UNX2'] = df.Date.astype('int64')//10**9
# dfA_M['UNX_time'] = dfA_M.DateTime.astype('int64')//10**9
dfA_M['UNX_time'] = dfA_M.DateTime.astype('int64') //10**9

<ipython-input-18-ce1795b96345>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfA_M['UNX_time'] = dfA_M.DateTime.astype('int64') //10**9


In [212]:
dfA_M.head()

order_created_datetime brand_name  \
52465  2015-09-13 04:00:11.000000          A   
25870  2015-09-13 09:11:50.000000          A   
25804  2015-09-13 09:12:49.000000          A   
23945  2015-09-13 09:53:28.000000          A   
23701  2015-09-13 10:00:37.000000          A   

                                rest_id             batched_with_order_id  \
52465  6c1ad5c66540b8c174ff7d9307027299  d0febfc6653fa848cf0a9fadf98261dd   
25870  8aec6c26cf9db4db10ed06e5e39681ed  04c7d89ab6b8ff72fb5868413602ddef   
25804  8aec6c26cf9db4db10ed06e5e39681ed  5ae2e25193da2000e07dc2051b5ca657   
23945  707f1da36335ef7234721b49991d39c3  83044987a923b13056db64220dd548c5   
23701  707f1da36335ef7234721b49991d39c3  287564b3ea79fef51a657c6a4237deb5   

                               order_id  first_in_multiorder_flg  \
52465  931ad84269827a4c27ce66734460cdf7                      0.0   
25870  5ae2e25193da2000e07dc2051b5ca657                      1.0   
25804  04c7d89ab6b8ff72fb5868413602ddef                      0.0   
23945  287564b3ea79fef51a657c6a4237deb5                      0.0   
23701  83044987a923b13056db64220dd548c5                      1.0   

                             courier_id  order_items_cnt  \
52465  a973a8d1969c72a9eb8147d995e519bf                7   
25870  ebe67e247939ab69f7734ba1557b763d                2   
25804  ebe67e247939ab69f7734ba1557b763d               14   
23945  cd30a21c6bbbcc784de70aa8627be584                8   
23701  cd30a21c6bbbcc784de70aa8627be584                5   

                                   city            DateTime    UNX_time  
52465  e6705560085199e9674cd721ee844d0d 2015-09-13 04:00:11  1442116811  
25870  e6705560085199e9674cd721ee844d0d 2015-09-13 09:11:50  1442135510  
25804  e6705560085199e9674cd721ee844d0d 2015-09-13 09:12:49  1442135569  
23945  e6705560085199e9674cd721ee844d0d 2015-09-13 09:53:28  1442138008  
23701  e6705560085199e9674cd721ee844d0d 2015-09-13 10:00:37  1442138437

Закрепяем i (номер заказа), начинаем перебирать j(номер заказа который прибавили). как только находим совпадения в столбцах, вычитаем из более раннего более поздний. 


In [19]:
delta_time = []
for i in range(len(dfA_M)):
    for j in range(len(dfA_M)):
        if (dfA_M.order_id.iloc[j] == dfA_M.batched_with_order_id.iloc[i]):
            if (dfA_M.first_in_multiorder_flg.iloc[j] == 1.0):
                delta_time.append(abs(dfA_M.UNX_time.iloc[j] - dfA_M.UNX_time.iloc[i]))

delta_time - список с разницами во времени мультизаказов выраженный в секундах. 

In [78]:
delta_time = delta_time[1:] # удалил самый первый элемент, я хз что там такое, но он портит всю картину. 
delta_time


[59,
 429,
 109,
 281,
 505,
 225,
 201,
 126,
 329,
 331,
 358,
 287,
 99,
 736,
 410,
 797,
 340,
 336,
 293,
 180,
 211,
 448,
 278,
 99,
 503,
 551,
 155,
 550,
 560,
 82,
 443,
 402,
 239,
 436,
 579,
 38,
 52,
 364,
 1275,
 46,
 804,
 313,
 547,
 30,
 72,
 638,
 263,
 589,
 15,
 168,
 617,
 788,
 98,
 735,
 487,
 140,
 678,
 292,
 406,
 159,
 39,
 332,
 644,
 567,
 474,
 44,
 424,
 694,
 34,
 576,
 1125,
 93,
 1028,
 173,
 149,
 830,
 959,
 1357,
 194,
 1222,
 10,
 1347,
 1031,
 663,
 703,
 541,
 79,
 6,
 285,
 814,
 467,
 1148,
 912,
 72,
 479,
 276,
 248,
 164,
 949,
 642,
 166,
 368,
 180,
 194,
 148,
 265,
 477,
 113,
 548,
 241,
 42,
 315,
 630,
 363,
 833,
 356,
 481,
 516,
 724,
 618,
 607,
 661,
 810,
 537,
 124,
 194,
 868,
 120,
 12,
 252,
 650,
 272,
 653,
 535,
 104,
 77,
 224,
 237,
 330,
 374,
 1036,
 163,
 258,
 155,
 165,
 1010,
 365,
 481,
 18,
 603,
 515,
 415,
 855,
 699,
 603,
 306,
 2072,
 75,
 460,
 518,
 470,
 599,
 266,
 172,
 768,
 333,
 272,
 108,
 60,
 

t_multi - максимальное время за который делают мультизаказ. Если время между заказами в выборке В превышает данное значение, мульти заказ создаваться не будет. Будем его использовать как критерий 

In [20]:
t_multi = max(delta_time)
t_multi

31188

Создаём ДатаФрейм с заказами рестранов В, сразу добавляем столбцы с Датой и временем в формате Дата и столбец с датой в формате UNIX

In [79]:
dfB['DateTime'] = pd.to_datetime(dfB['order_created_datetime'], format='%Y-%m-%d %H:%M:%S.%f') #%S%d%b%Y:%H:%M:%S.%f')
dfB['UNX_time'] = dfB.DateTime.astype('int64') //10**9
dfB = dfB.sort_values(by=['DateTime'])
dfB.head()

<ipython-input-79-c306c3a4c20d>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfB['DateTime'] = pd.to_datetime(dfB['order_created_datetime'], format='%Y-%m-%d %H:%M:%S.%f') #%S%d%b%Y:%H:%M:%S.%f')
<ipython-input-79-c306c3a4c20d>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dfB['UNX_time'] = dfB.DateTime.astype('int64') //10**9


order_created_datetime brand_name  \
41675  2015-09-13 00:00:44.000000          B   
41533  2015-09-13 00:00:53.000000          B   
61228  2015-09-13 00:01:47.000000          B   
41753  2015-09-13 00:03:43.000000          B   
50735  2015-09-13 00:04:21.000000          B   

                                rest_id batched_with_order_id  \
41675  69a08198a23fac8f83359e2ec23a49b6                   NaN   
41533  69a08198a23fac8f83359e2ec23a49b6                   NaN   
61228  938cc47b8207a782316f907be53aab3c                   NaN   
41753  69a08198a23fac8f83359e2ec23a49b6                   NaN   
50735  d3fba5f2e58b2a4f4dc4bf94d6cc3e49                   NaN   

                               order_id  first_in_multiorder_flg  \
41675  d9f50562ab79c432e64413fcee7aeb9c                      NaN   
41533  5cc2b1699fdbeee506fb770ff77c4ff9                      NaN   
61228  b43251899246c62369370f143f2e73b2                      NaN   
41753  943eee04ddeee9b44635947d7c0a8f41                      NaN   
50735  75863a090739e1b8c50747bea638639a                      NaN   

                             courier_id  order_items_cnt  \
41675  8ec5e3cc3bc59b70a0c8863b0355cc92                3   
41533  50c20cfed81be8a8fbcd8dc5b2e2ab76                7   
61228  4b6de04194b5a6f6ebdf6e09458eab6f                1   
41753  768f844a816657e7533495bd01bc2a80                3   
50735  b30f23a13ea180021abfcb8f98d126cc                1   

                                   city            DateTime    UNX_time  
41675  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:44  1442102444  
41533  e6705560085199e9674cd721ee844d0d 2015-09-13 00:00:53  1442102453  
61228  e6705560085199e9674cd721ee844d0d 2015-09-13 00:01:47  1442102507  
41753  e6705560085199e9674cd721ee844d0d 2015-09-13 00:03:43  1442102623  
50735  e6705560085199e9674cd721ee844d0d 2015-09-13 00:04:21  1442102661

In [80]:
multi_count_1 = 0
for i in range(len(dfB)-1):
    '''Перебираем строки и проверяем если номера заказов в строке и строке +1 совпадают, то сравниваем время, 
    если время в менее максимального времени в Мультизаказах ресторанов А, добавляем +1 в счeтчик'''
    if dfB.rest_id.iloc[i] == dfB.rest_id.iloc[i+1] and (dfB.UNX_time.iloc[i] - dfB.UNX_time.iloc[i+1]) < t_multi:
        multi_count += 1
print(multi_count)
            

251


In [85]:
multi_count_2 = 0
rest = 0
        
def dfB_new():
    '''Создаём функцию, которая создаёт ДатаФреймы по всем уникальным rest_id и сотрирует по дате в формате Юникс'''
    for rest in dfB.rest_id.unique():
        df_new = dfB[dfB.rest_id == rest]
        df_new = df_new.sort_values(by=['UNX_time'])
        return df_new
         
for time in range(len(dfB_new())-1):
    '''Считаем сколько раз встречается разница во времени между заказми одного rest_id, 
    что бы собрать мультизаказ по тем же условиям по которым собирались мульти заказы в выборке А'''
    if (dfB_new().UNX_time.iloc[time] - dfB_new().UNX_time.iloc[time+1]) < t_multi:
        multi_count_2 += 1
print(multi_count_2)

276


In [88]:
result_multiorder = (multi_count_2 * 2) / len(dfB)
print("Процент двойных заказов от общего числа заказов", round(result_multiorder*100, 2),'%')

Процент двойных заказов от общего числа заказов 0.97 %
